File: 113-path-sum-ii/path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>>res;
    void dfs(TreeNode *root, vector<int>target, int targetSum, int sum){
        if(root==NULL) return;
        target.push_back(root->val);
        sum+=root->val;
        if(root->left==NULL && root->right==NULL && targetSum==sum){
            res.push_back(target);
        }

        dfs(root->left, target, targetSum, sum);
        dfs(root->right, target, targetSum, sum);
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        res.clear();
        dfs(root,{},targetSum,0);
        return res;
    }
};

File: 113-path-sum-ii/path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>>res=new ArrayList<>();
    public void dfs(TreeNode root, int sum, List<Integer>row, int targetSum){
        if(root==null) return;
        sum+=root.val;
        row.add(root.val);
        if(root.left==null && root.right==null && sum==targetSum){
            res.add(new ArrayList<>(row));
        }
        dfs(root.left,sum,row,targetSum);
        dfs(root.right,sum,row,targetSum);
        row.remove(row.size() - 1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if(root==null) return res;
        int sum=0;
        List<Integer>row=new ArrayList<>();
        dfs(root,sum,row,targetSum);
        return res;
    }
}